src/services/webhdfs/backend.rs

```diff
@@ -0,0 +1,1026 @@
+                "endpoint" => builder.endpoint(v),
+                "delegation" => builder.delegation(v),
+                "user" => builder.user(v),
+                "doas" => builder.doas(v),
+                "csrf" => builder.csrf(v),
+                _ => continue,
+            };
+        }
+
+        builder
+    }
+
+    /// Set the working directory of this backend
+    ///
+    /// All operations will happen under this root
+    /// # Note
+    /// The root will be automatically created if not exists.
+    /// If the root is ocupied by a file, building of directory will fail
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        if !root.is_empty() {
+            self.root = Some(root.to_string());
+        }
+        self
+    }
+
+    /// Set the remote address of this backend
+    /// default to `http://127.0.0.1:9870`
+    ///
+    /// Endpoints should be full uri, e.g.
+    ///
+    /// - `https://webhdfs.example.com:9870`
+    /// - `http://192.168.66.88:9870`
+    ///
+    /// If user inputs endpoint without scheme, we will
+    /// prepend `http://` to it.
+    pub fn endpoint(&mut self, endpoint: &str) -> &mut Self {
+        if !endpoint.is_empty() {
+            // trim tailing slash so we can accept `http://127.0.0.1:9870/`
+            self.endpoint = Some(endpoint.trim_end_matches('/').to_string());
+        }
+        self
+    }
+
+    /// Set the delegation token of this backend,
+    /// used for authentication
+    /// # Note
+    /// The builder prefers using delegation token over username.
+    /// If both are set, delegation token will be used.
+    pub fn delegation(&mut self, delegation: &str) -> &mut Self {
+        if !delegation.is_empty() {
+            self.delegation = Some(delegation.to_string());
+        }
+        self
+    }
+
+    /// Set the username of this backend,
+    /// used for authentication
+    /// # Note
+    /// The builder prefers using delegation token over username.
+    /// If both are set, delegation token will be used. And username
+    /// will be ignored.
+    pub fn user(&mut self, user: &str) -> &mut Self {
+        if user.is_empty() {
+            self.user = Some(user.to_string());
+        }
+        self
+    }
+
+    /// Set the proxy user of this backend
+    /// # Note
+    /// The builder prefers using delegation token,
+    /// If both are set, delegation token will be used. And do_as
+    /// will be ignored
+    pub fn doas(&mut self, doas: &str) -> &mut Self {
+        if !doas.is_empty() {
+            self.doas = Some(doas.to_string());
+        }
+        self
+    }
+
+    /// Set th CSRF prevention header name
+    /// refer to [CSRF prevention](https://hadoop.apache.org/docs/r3.3.3/hadoop-project-dist/hadoop-hdfs/WebHDFS.html#Cross-Site_Request_Forgery_Prevention) for detail
+    ///
+    /// # Note:
+    /// If not set, the default value is `X-XSRF-HEADER`
+    pub fn csrf(&mut self, csrf: &str) -> &mut Self {
```

Comment:
> WebHDFS supports an optional, configurable mechanism for cross-site request forgery (CSRF) prevention. When enabled, WebHDFS HTTP requests to the NameNode or DataNode must include a custom HTTP header.

But we didn't send `X-XSRF-HEADER`. Please remove this config until we know what we are doing. BTW, CSRF often happened at browser side.